### PR TITLE
[candi] build kubernetes components as binaries

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
## Description
1. Kubelet, kubectl and kubeadm binaries build as static and installs into the `/opt/deckhouse/bin` directory.
2. Containerd and runc binaries build as static and installs into the `/opt/deckhouse/bin` directory.
3. All deckhouse system scripts moved to the `/opt/deckhouse/bin` directory.

## Why do we need it, and what problem does it solve?
Delivering Kubernetes and containerd as static binaries simplifies the install process on different distros.
Moving all node binaries installed by Deckhouse to a special directory resolve the problem with some distro security scanners which detects changes in system directories.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
```changes
section: candi
type: chore
summary: Deliver Kubernetes node components as binaries.
impact: All the Kubernetes control plane and node components will restart.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by RomanenkoDenys** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.25 4806987080 4806987080-1-con-1-25 /sys/deckhouse-oss/install:pr4458

**Comment by RomanenkoDenys** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 4806984318 4806984318-1-con-1-25 /sys/deckhouse-oss/install:pr4458

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by RomanenkoDenys** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 4807495708 4807495708-1-con-1-25 /sys/deckhouse-oss/install:pr4458 centos@3.77.242.88

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by RomanenkoDenys** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 4808057314 4808057314-1-con-1-25 /sys/deckhouse-oss/install:pr4458 centos@18.192.189.30

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by RomanenkoDenys** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 4808057314 4808057314-1-con-1-25 /sys/deckhouse-oss/install:pr4458

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by RomanenkoDenys** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 4808347642 4808347642-1-con-1-25 /sys/deckhouse-oss/install:pr4458 centos@3.77.243.228

**Comment by RomanenkoDenys** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.25 4808533252 4808533252-1-con-1-25 /sys/deckhouse-oss/install:pr4458 user@34.159.183.161

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by RomanenkoDenys** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.25 4808533252 4808533252-1-con-1-25 /sys/deckhouse-oss/install:pr4458 user@34.159.183.161

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by RomanenkoDenys** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.25 4809029691 4809029691-1-con-1-25 /sys/deckhouse-oss/install:pr4458 user@35.242.196.193

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by z9r5** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.25 4806987080 4806987080-1-con-1-25 /sys/deckhouse-oss/install:pr4458 user@35.246.159.249

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

File: modules/007-registrypackages/images/kubelet/scripts/uninstall [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# Copyright 2022 Flant JSC
+# Copyright 2021 Flant JSC
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply modules/007-registrypackages/images/kubelet/scripts/uninstall:2]: fixed to 2023 in all scripts

File: modules/000-common/images/kubernetes/werf.inc.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  - for patchfile in /patches/kube-apiserver/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
+  - for patchfile in /patches/kube-controller-manager/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
+  - for patchfile in /patches/kubeadm/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
+  - make all WHAT=cmd/kube-apiserver
+  - make all WHAT=cmd/kube-controller-manager
+  - make all WHAT=cmd/kube-scheduler
+  - make all WHAT=cmd/kube-proxy
+  - make all WHAT=cmd/kubelet GOLDFLAGS="-s -w -extldflags \"-static\""
+  - make all WHAT=cmd/kubectl
+  - make all WHAT=cmd/kubeadm GOLDFLAGS="-s -w -extldflags \"-static\""
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply modules/000-common/images/kubernetes/werf.inc.yaml:50]: others build static by default but i added GOLDFLAGS for similarity

File: modules/000-common/images/kubernetes/werf.inc.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    KUBE_GIT_VERSION='v{{ printf "%s.%s" $version $patch }}'
+    KUBE_GIT_MAJOR='{{ $v.Major }}'
+    KUBE_GIT_MINOR='{{ $v.Minor }}'
+    KUBE_GIT_COMMIT='0000000000000000000000000000000000000000'
+    KUBE_GIT_TREE_STATE='archive'
+    EOF
+  - export KUBE_GIT_VERSION_FILE=.kube-version
+  - for patchfile in /patches/kube-apiserver/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
+  - for patchfile in /patches/kube-controller-manager/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
+  - for patchfile in /patches/kubeadm/{{ $version }}/*.patch ; do patch -p1 < ${patchfile}; done
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply modules/000-common/images/kubernetes/werf.inc.yaml]: done

File: candi/bashible/common-steps/all/001_add_deckhouse_path.sh.tpl [hunk 4]
```diff
@@ -13,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-set -Eeo pipefail
-dpkg -i -E cri-tools_amd64.deb kubeadm_amd64.deb
-apt-mark hold kubeadm cri-tools
+#!/bin/bash
+
```
Comment (apolovov) [inline candi/bashible/common-steps/all/001_add_deckhouse_path.sh.tpl:16]:
```suggestion
```

File: candi/bashible/bundles/centos/bootstrap.sh.tpl [hunk 5]
```diff
@@ -164,7 +164,7 @@ bb-rp-install() {
 {{- /*
 # IMPORTANT !!! Do not remove this line, because in Centos/Redhat when dhctl bootstraps the cluster /usr/local/bin not in PATH.
 */}}
```
Comment (apolovov) [inline candi/bashible/bundles/centos/bootstrap.sh.tpl:166]:
```suggestion

```

File: candi/bashible/common-steps/all/034_install_kubelet_and_his_friends.sh.tpl [hunk 6]
```diff
@@ -13,16 +13,16 @@
 # limitations under the License.
 
 {{- $kubernetesVersion := printf "%s%s" (.kubernetesVersion | toString) (index .k8s .kubernetesVersion "patch" | toString) | replace "." "" }}
-{{- $kubernetesCniVersion := index .k8s .kubernetesVersion "cniVersion" | toString | replace "." "" }}
+{{- $kubernetesCniVersion := "1.2.0" | replace "." "" }}
```
Comment (apolovov) [inline candi/bashible/common-steps/all/034_install_kubelet_and_his_friends.sh.tpl:16]: Why hardcoded?

File: candi/bashible/common-steps/all/063_install_kubelet_forker.sh.tpl [hunk 7]
```diff
@@ -14,13 +14,13 @@
 
 #!/bin/bash
```
Comment (apolovov) [inline candi/bashible/common-steps/all/063_install_kubelet_forker.sh.tpl:16]:
```suggestion
```

File: candi/bashible/common-steps/cluster-bootstrap/035_install_kubeadm.sh.tpl [hunk 8]
```diff
@@ -11,9 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
 {{- $kubernetesVersion := printf "%s%s" (.kubernetesVersion | toString) (index .k8s .kubernetesVersion "patch" | toString) | replace "." "" }}
 {{- $kubernetesMajorVersion := .kubernetesVersion | toString | replace "." "" }}
-{{- $kubernetesCniVersion := index .k8s .kubernetesVersion "cniVersion" | toString | replace "." "" }}
-
-bb-rp-install "kubeadm:{{ index .images.registrypackages (printf "kubeadmCentos%s" $kubernetesVersion) }}" "kubelet:{{ index .images.registrypackages (printf "kubeletCentos%s" $kubernetesVersion) }}" "kubectl:{{ index .images.registrypackages (printf "kubectlCentos%s" $kubernetesVersion) }}" "crictl:{{ index .images.registrypackages (printf "crictl%s" $kubernetesMajorVersion) }}" "kubernetes-cni:{{ index .images.registrypackages (printf "kubernetesCniCentos%s" $kubernetesCniVersion) }}"
+{{- $kubernetesCniVersion := "1.2.0" | replace "." "" }}
```
Comment (apolovov) [inline candi/bashible/common-steps/cluster-bootstrap/035_install_kubeadm.sh.tpl:16]: Hardcode again.

File: candi/bashible/common-steps/node-group/031_install_containerd.sh.tpl [hunk 9]
```diff
@@ -14,19 +14,17 @@
...
   systemctl daemon-reload
   systemctl enable containerd.service
+
 {{ if ne .runType "ImageBuilding" -}}
   bb-flag-set containerd-need-restart
 {{- end }}
 }
 
-if bb-yum-package? docker-ce; then
+if bb-apt-package? docker-ce || bb-apt-package? docker.io; then
```
Comment (apolovov) [inline candi/bashible/common-steps/node-group/031_install_containerd.sh.tpl:27]: Need to discuss, i didn' get how it used to work in common-steps outside centos bundle and i don't get how it works now in centos node.

File: candi/tools/update_kubernetes_patchversions.sh [hunk 10]
```diff
@@ -27,19 +27,7 @@ function check_requirements() {
 # Updates version map with new patchversion
 # update_version_map VERSION PATCH
 function update_version_map() {
-  local NEW_DIGEST
   yq -i e ".k8s.\"${1}\".patch = ${2}" ../version_map.yml
-  # Kube-proxy
-  if yq -e e "select(.k8s.\"${1}\".controlPlane | has(\"kubeProxy\"))" ../version_map.yml >/dev/null 2>/dev/null; then
-    NEW_DIGEST="$(crane digest "registry.k8s.io/kube-proxy:v${1}.${2}")"
```
Comment (apolovov) [inline candi/tools/update_kubernetes_patchversions.sh:34]: to discuss

**Review by apolovov — CHANGES_REQUESTED** [review]

File: candi/bashible/common-steps/node-group/031_install_containerd.sh.tpl (hunk 9 above)
Comment (RomanenkoDenys) [reply candi/bashible/common-steps/node-group/031_install_containerd.sh.tpl:27]: good point, i'll restore original scripts

File: candi/tools/update_kubernetes_patchversions.sh (hunk 10 above)
Comment (RomanenkoDenys) [reply candi/tools/update_kubernetes_patchversions.sh:34]: we build kube-proxy from source, so we do not need to pass hash sum to version map

**Review by apolovov — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Pull request merged**